File: exercises/easy-level/14.cpp
// 14. Longest Common Prefix
#include <iostream>
#include <vector>

using namespace std;

string longestCommonPrefix(vector<string>& strs) {
    string min = strs[0];
    for (string i : strs)
        if (i.size() < min.size())
            min = i;

    if (min.size() == 0) return "";

    int right = min.size();
    for (string str : strs) {
        for (int i = 0; i < right; i++) {
            if (str[i] != min[i]) {
                right = i;
                break;
            }
        }
    }

    return min.substr(0, right);
}

int main()
{
  vector<string> test1, test2;
  test1 = {"flower","flow","flight"};
  test2 = {"dog","racecar","car"};
  
  cout << longestCommonPrefix(test1) << endl;
  cout << longestCommonPrefix(test2) << endl;
  return 0;
}


/*
Example 1:

Input: strs = ["flower","flow","flight"]
Output: "fl"


Example 2:

Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
 

Constraints:

1 <= strs.length <= 200
0 <= strs[i].length <= 200
strs[i] consists of only lowercase English letters if it is non-empty.
*/

File: easy-level-golang/14/14.go
package main

import (
	"fmt"
	"strings"
)

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	shortStr := strs[0]
	for _, st := range strs {
		if len(st) < len(shortStr) {
			shortStr = st
		}
	}

	count := 0
	for count < len(strs) && len(shortStr) > 0 {
		if !strings.HasPrefix(strs[count], shortStr) {
			shortStr = shortStr[:len(shortStr)-1]
			continue
		}
		count++
	}

	return shortStr
}

func main() {
	fmt.Println(longestCommonPrefix([]string{"flower", "flow", "flight"}))
	fmt.Println(longestCommonPrefix([]string{"dog", "racecar", "car"}))
}
